pandas/tests/indexes/ranges/test_range.py

```diff
@@ -608,6 +608,16 @@ def test_range_index_rsub_by_const(self):
         tm.assert_index_equal(result, expected)
 
 
+def test_reindex_returns_rangeindex():
+    ri = RangeIndex(2, name="foo")
+    result, result_indexer = ri.reindex([1, 2, 3])
+    expected = RangeIndex(1, 4, name="foo")
+    tm.assert_index_equal(result, expected, exact=True)
+
+    expected_indexer = np.array([1, -1, -1], dtype=np.intp)
+    tm.assert_numpy_array_equal(result_indexer, expected_indexer)
```

Comment:
Would it make sense to add a test case when RangeIndex can't be returned? Like, if we call `ri.reindex([1, 3, 9])` my expectation is that it should still return `Index`, which I guess it does, but I wonder whether your changes can make it raise instead. I think it'd be good to have a test for that, and I couldn't find one already in the test suite.